pkg/controller/direct/gkehub/featuremembership_controller.go

```diff
@@ -0,0 +1,213 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gkehub
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	featureapi "google.golang.org/api/gkehub/v1beta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/pkg/clients/generated/apis/gkehub/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+)
+
+const ctrlName = "gkehubfeaturemembership-controller"
+
+// AddGkeHubController creates a new controller and adds it to the Manager.
+// The Manager will set fields on the Controller and start it when the Manager is started.
+func AddGkeHubController(mgr manager.Manager, config *controller.Config, opts directbase.Deps) error {
+	gvk := krm.GKEHubFeatureMembershipGVK
+
+	gcpClient, err := newGCPClient(config)
+	if err != nil {
+		return err
+	}
+	m := &gkeHubModel{gcpClient: gcpClient}
+	return directbase.Add(mgr, gvk, m, opts)
+}
+
+type gkeHubModel struct {
+	gcpClient *gcpClient
+}
+
+// model implements the Model interface.
+var _ directbase.Model = &gkeHubModel{}
+
+type gkeHubAdapter struct {
+	membershipID string
+	featureID    string
+	projectID    string
+	location     string
+
+	desired *featureapi.MembershipFeatureSpec
+	actual  *featureapi.Feature
+
+	featureClient *featureapi.ProjectsLocationsFeaturesService
+}
+
+var _ directbase.Adapter = &gkeHubAdapter{}
+
+// AdapterForObject implements the Model interface.
+func (m *gkeHubModel) AdapterForObject(ctx context.Context, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	projectsLocationsFeaturesService, err := m.gcpClient.newProjectsLocationsFeaturesService(ctx)
+	if err != nil {
+		return nil, err
+	}
+	obj := &krm.GKEHubFeatureMembership{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+	// TODO(ziyue): resolve the external ref
+	projectID := obj.Spec.ProjectRef.Name
+	if projectID == "" {
+		return nil, fmt.Errorf("cannot resolve projectID")
+	}
+	mID, err := fullyQualifiedNameForMembership(*obj)
+	if err != nil {
+		return nil, err
+	}
+	fID, err := fullyQualifiedNameForFeature(*obj)
+	if err != nil {
+		return nil, err
+	}
+	apiObj, err := featureMembershipSpecKRMtoMembershipFeatureSpecAPI(&obj.Spec)
+	if err != nil {
+		return nil, err
+	}
+	return &gkeHubAdapter{
+		membershipID:  mID,
+		featureID:     fID,
+		projectID:     "projects/" + projectID,
```

Comment:
Super-nit (just a tip): personally I find it easier if a field named `projectID` is just the ID, so I would name this as `projectFQN` or `projectLink` or `projectPath` or something like that.  Or, more probably, I would just store the projectID and add `projects/` when I combine it.

We'll probably end up with something strong-typed anyway once we define shared functions around resolving projects, so not a big deal!

(FQN = Fully Qualified Name)

Reply:
Remove the "projects/" prefix in the projectID